src/net/p2p/broadcaster.cpp

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright (c) [2023-2024] [Sparq Network]
+
+This software is distributed under the MIT License.
+See the LICENSE.txt file in the project root for more information.
+*/
+
+#include "broadcaster.h"
+#include "managernormal.h"
+#include "../../core/blockchain.h"
+
+// FIXME/TODO: check which ones are redundant
+#include "../core/rdpos.h"
+#include "../core/storage.h"
+#include "../core/state.h"
+#include "nodeconns.h"
+
+namespace P2P {
+
+  const Options& Broadcaster::getOptions() { return manager_.getOptions(); }
+
+  void Broadcaster::broadcastMessage(const std::shared_ptr<const Message> message, const std::optional<NodeID>& originalSender) {
+    this->manager_.sendMessageToAll(message, originalSender);
+  }
+
+  void Broadcaster::handleTxValidatorBroadcast(
+    const NodeID &nodeId, const std::shared_ptr<const Message>& message
+  ) {
+    // FIXME/REVIEW: This has a rebroadcasting condition. We need to ensure this
+    //   condition is sufficient to detect "duplicate broadcasts". If it is, then
+    //   so far we won't be needing a rebroadcast detection mechanism that's internal
+    //   to the Broadcaster which is based on message IDs.
+
+    try {
+      auto tx = BroadcastDecoder::broadcastValidatorTx(*message, getOptions().getChainID());
+      if (this->state_.addValidatorTx(tx)) this->broadcastMessage(message, nodeId);
+    } catch (std::exception const& ex) {
+      throw DynamicException("Invalid txValidatorBroadcast (" + std::string(ex.what()) + ")");
+    }
+  }
+
+  void Broadcaster::handleTxBroadcast(
+    const NodeID &nodeId, const std::shared_ptr<const Message>& message
+  ) {
+    // FIXME/REVIEW: This has a rebroadcasting condition. We need to ensure this
+    //   condition is sufficient to detect "duplicate broadcasts". If it is, then
+    //   so far we won't be needing a rebroadcast detection mechanism that's internal
+    //   to the Broadcaster which is based on message IDs.
+
+    try {
+      auto tx = BroadcastDecoder::broadcastTx(*message, getOptions().getChainID());
+      if (!this->state_.addTx(std::move(tx))) this->broadcastMessage(message, nodeId);
```

Comment:
Same as the previous comment, but we're looking at `State::validateTransactionInternal()` instead. This tx will only be broadcast if ALL of those match:

* A's tx was already in B's mempool
* A's tx has a valid balance and nonce
* The tx sender has enough balance to send it

Like the previous comment, until we introduce stricter tx consensus rules (or someone else objects), this seems to suffice IMO. The burden of "broadcasting or not" falls on the consensus rules' shoulders as it is coded right now.

Plus, we're checking an enum that should be 0 (`TxInvalid::NotInvalid`, thus the NOT "!" check). Honestly I would just suggest (if possible) to check against the enum directly instead of relying on its value, just so it's a bit more readable. Example:

```suggestion
      if (this->state_.addTx(std::move(tx)) == TxInvalid::NotInvalid) this->broadcastMessage(message, nodeId);
```

Reply:
I have changed the `TxInvalid` enum type name to `TxStatus`, and I have broken `NotInvalid` into `ValidNew` and `ValidExisting`. This adds zero code to State, just two `teturns` that were returning the same enum value will now return two different values. This allows Broadcaster to differentiate precisely when it should actually broadcast it (i.e. `ValidNew`).